### PR TITLE
Fix Content-Length requirement, header updates in payload signer, http fallback to signed

### PR DESCRIPTION
**Pull request by haydenbaker**
## Motivation and Context
- Content-Length should be pre-calculated for aws-chunked content-encoding and set, so that the http-client does not need to buffer the payload in memory to calculate it for us
- Updates to headers on the request by the payload signer must happen before request is signed, otherwise the canonical request that is signed will be incorrect
- Unsigned-payload with HTTP should fall back to signed-payload, just as the old signers do

## Testing
- Run new and existing tests, make sure pass

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/http-auth-aws/src/test/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSignerTest.java [hunk 1]
```diff
@@ -291,4 +292,19 @@ public void sign_WithEventStreamContentTypeAndUnsignedPayload_Throws() {
 
         assertThrows(UnsupportedOperationException.class, () -> signer.signAsync(request));
     }
+
+    @Test
+    public void sign_WithPayloadSigningFalseAndHttp_FallsBackToPayloadSigning() {
```
[the remark this answers is not in the document]
Comment (haydenbaker) [reply core/http-auth-aws/src/test/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSignerTest.java:298]: Sure

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/AwsChunkedV4aPayloadSigner.java [hunk 2]
```diff
@@ -101,49 +102,152 @@ public ContentStreamProvider sign(ContentStreamProvider payload, V4aContext v4aC
...
+                break;
+            }
+            case STREAMING_UNSIGNED_PAYLOAD_TRAILER:
+                if (checksumAlgorithm != null) {
+                    encodedContentLength += calculateChecksumTrailerLength(checksumHeaderName(checksumAlgorithm));
+                }
+                encodedContentLength += calculateChunksLength(contentLength, 0);
+                break;
+            case STREAMING_ECDSA_SIGNED_PAYLOAD_TRAILER: {
+                long extensionsLength = 161; // ;chunk-signature:<sigv4a-ecsda hex signature, 64 bytes>
```
Comment (haydenbaker) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/AwsChunkedV4aPayloadSigner.java:127]: Fix comment

File: core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/util/SignerUtils.java [hunk 3]
```diff
@@ -252,4 +254,21 @@ public static byte[] hash(byte[] data) {
     public static byte[] hash(String text) {
         return hash(text.getBytes(StandardCharsets.UTF_8));
     }
+
+    private static int readAll(InputStream inputStream) {
```
Comment (cenedhryn) [inline core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/util/SignerUtils.java:258]: Add comment why it's safe to use this method on the supplied stream

**Review by cenedhryn — APPROVED** [review]
Approving - didn't look at naming etc, which could be refactored later.

**Pull request merged**